:sparkles: Add support for fetching multiple kubeconfig

valaparthvi opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:
This PR adds support to fetch a different type of Kubeconfig. `clusterctl get kubeconfig <name>` now fetches the user specific kubeconfig by default, and if unavailable, it falls back to fetching the system specific kubeconfig. 
It also adds a new flag `--user` which is true by default, and can be set to false to receive system specific kubeconfig.

See #3661 for more details.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #3661

valaparthvi commented:
@richardcase Do you mind taking a look at this (whenever you find the time, of course)? I am not very sure about the documentation and comments, and I feel as if I might have missed out on test cases, but I didn't find any integration test that would need to be modified.

sbueringer commented:
/ok-to-test

richardcase commented:
@valaparthvi - thanks for working on this :+1: I think we should be aware of [this](https://github.com/kubernetes-sigs/cluster-api/issues/3661#issuecomment-694747492) comment from @fabriziopandini where we should default to returning the "user" kubeconfig if there is one and if there isn't then fallback to the current one as per the current behaviour of clusterctl.

What this means is that we should look for a secret with name that uses this format first `[cluster-name]-user-kubeconfig`. If that's not found, then fallback to looking for the secret with `[cluster-name]-kubeconfig` (which is the current behaviour). 

We could add a flag that allows us to get the "system" kubeconfig instead from clusterctl instead of the user one. So something like: `clusterctl get kubeconfig --system`.

If you want to jump on a call any time feel free to ping me on slack.

valaparthvi commented:
/test pull-cluster-api-test-main
>
```{Failed  === RUN   TestClusterClassWebhook_Delete_MultipleExistingClusters
    clusterclass_test.go:460: 
        Expected
            <*errors.StatusError | 0xc0012c5720>: {
                ErrStatus: {
                    TypeMeta: {Kind: "", APIVersion: ""},
                    ListMeta: {
                        SelfLink: "",
                        ResourceVersion: "",
                        Continue: "",
                        RemainingItemCount: nil,
                    },
                    Status: "Failure",
                    Message: "admission webhook \"validation.clusterclass.cluster.x-k8s.io\" denied the request: ClusterClass.cluster.x-k8s.io \"class4\" is forbidden: ClusterClass cannot be deleted because it is used by 1 Cluster(s)",
                    Reason: "Forbidden",
                    Details: {
                        Name: "class4",
                        Group: "cluster.x-k8s.io",
                        Kind: "ClusterClass",
                        UID: "",
                        Causes: nil,
                        RetryAfterSeconds: 0,
                    },
                    Code: 403,
                },
            }
        to be nil
--- FAIL: TestClusterClassWebhook_Delete_MultipleExistingClusters (0.51s)
}
```

richardcase commented on cmd/clusterctl/cmd/get_kubeconfig.go:
```diff
@@ -59,6 +63,8 @@ func init() {
 	getKubeconfigCmd.Flags().StringVar(&gk.kubeconfigContext, "kubeconfig-context", "",
 		"Context to be used within the kubeconfig file. If empty, current context will be used.")
+	getKubeconfigCmd.Flags().BoolVar(&gk.userKubeconfig, "system", true,
+		"If set to false, returns the kubeconfig of the system, else returns the user specific kubeconfig.")
```
We should probably define what we mean by **system** kubeconfig somewhere. Perhaps in docs.

richardcase replied:
What we are referring to here as the `system kubeconfig` is the kubeconfig used by CAPI to connect to child clusters.

valaparthvi replied:
Yes. The one used by CAPI, or not-user kubeconfig is the system kubeconfig. The one currently provided by CAPI is system kubeconfig in this context.

richardcase commented on controllers/remote/cluster.go:
```diff
@@ -50,7 +50,7 @@ func NewClusterClient(ctx context.Context, sourceName string, c client.Client, c
 // RESTConfig returns a configuration instance to be used with a Kubernetes client.
 func RESTConfig(ctx context.Context, sourceName string, c client.Reader, cluster client.ObjectKey) (*restclient.Config, error) {
-	kubeConfig, err := kcfg.FromSecret(ctx, c, cluster)
+	kubeConfig, err := kcfg.FromSecret(ctx, c, cluster, true)
```
I think this should probably be `false` so that the behaviour stays the same (i.e. get the kubeconfig used by capi, a.k.a system kubeconfig)

richardcase commented on controlplane/kubeadm/internal/controllers/controller_test.go:
```diff
@@ -918,7 +918,7 @@ kubernetesVersion: metav1.16.1`,
 	g.Expect(s.Labels).To(Equal(expectedLabels))
 
-	k, err := kubeconfig.FromSecret(ctx, fakeClient, util.ObjectKey(cluster))
+	k, err := kubeconfig.FromSecret(ctx, fakeClient, util.ObjectKey(cluster), true)
```
Again, i think this should be `false` to preserve the existing behaviour. The only place where we should get the user kubeconfig by default is from the clusterctl command.

alexander-demichev replied:
+1 for keeping existing behaviour

richardcase commented on docs/book/src/clusterctl/commands/get-kubeconfig.md:
```diff
@@ -1,7 +1,8 @@
-This functionality is available in clusterctl v0.3.9 or newer.
+This command prints the kubeconfig of an existing workload cluster into stdout. This functionality is available in clusterctl v0.3.9 or newer.
+
+In a case where the provider generates a separate user specific Kubeconfig, clusterctl will return provider generated Kubeconfig by default.
```
```suggestion
In a case where the provider generates a separate user specific Kubeconfig, clusterctl will return the provider generated user Kubeconfig by default.
```

richardcase commented on util/secret/secret.go:
```diff
@@ -40,8 +41,17 @@ func GetFromNamespacedName(ctx context.Context, c client.Reader, clusterName cli
 	}
-
 	if err := c.Get(ctx, secretKey, secret); err != nil {
+		if purpose == UserKubeconfig {
```
I think this logic should sit in the kubeconfig package as this is just to get a secret with a specific purpose.

ykakarap commented on cmd/clusterctl/client/cluster/workload_cluster.go:
```diff
@@ -18,15 +18,17 @@ package cluster
 type WorkloadCluster interface {
 	// GetKubeconfig returns the kubeconfig of the workload cluster.
-	GetKubeconfig(workloadClusterName string, namespace string) (string, error)
+	GetKubeconfig(workloadClusterName string, namespace string, userKubeconfig bool) (string, error)
```
Changing the function signature here will cause a breaking API code change.

richardcase replied:
@valaparthvi could you introduce a new function? We can then refactor the existing one to call the new one.

fabriziopandini replied:
What about having a `GetUserKubeconfig` function, this will make things explicit while we can preserve the current API

ykakarap commented on util/kubeconfig/kubeconfig.go:
```diff
@@ -44,8 +44,12 @@ var (
 // FromSecret fetches the Kubeconfig for a Cluster.
-func FromSecret(ctx context.Context, c client.Reader, cluster client.ObjectKey) ([]byte, error) {
-	out, err := secret.Get(ctx, c, cluster, secret.Kubeconfig)
+func FromSecret(ctx context.Context, c client.Reader, cluster client.ObjectKey, userKubeconfig bool) ([]byte, error) {
```
Changing the function signature here will cause a breaking API code change.

richardcase replied:
@valaparthvi could you introduce a new function? We can then refactor the existing one to call the new one.

fabriziopandini replied:
same as above, (what about having `FromUserSecret`)

valaparthvi replied:
I can make that work.

ykakarap reviewed (commented):
Introducing a flag in `clusterctl get kubeconfig` to support multiple kubeconfigs looks like will introduce a few breaking API code changes in the clusterctl library. 

We have 2 approaches here: 
1) Introduce the flag, create the new functions with the required signatures, deprecate the original functions and provide support for fetching different kubeconfigs through `clusterctl get kubeconfig`.  (The functions that will need refactor are already marked in this review)
 
2) Keep inline with https://github.com/kubernetes-sigs/cluster-api/issues/3661#issuecomment-694747492 and lets just default to looking for `-user-kubeconfig` and fall back to looking for `-kubeconfig`. 
To do the above we can just simple alter the block here: https://github.dev/kubernetes-sigs/cluster-api/blob/70c9771dbccb2ecd24d3fc36a68d2e39d6ecd165/cmd/clusterctl/client/cluster/workload_cluster.go#L54-L57
to first look for the secret with `-user` and if not found fall back to the standard approach. 
We can do this by hacking the name of the workload cluster name to include `-user` and then simply use the `utilkubeconfig.FromSecret` function to get the kubeconfig.  

I am inclined to keeping this PR simple where it just defaults to `-user` (and falls back). 

We can approach the `clusterctl` flag along with the refactor in a separate PR as it probably needs more discussion. @fabriziopandini thoughts?

richardcase commented:
> We can approach the `clusterctl` flag along with the refactor in a separate PR as it probably needs more discussion. @fabriziopandini thoughts?

I think we need the flag to allow a user of clusterctl to explicitly get the "system" kubeconfig and not the user kubeconfig in the scenario where a provider publishes 2. But as @fabriziopandini mentioned in the comments the flag can default to getting the user kubeconfig (if there is one) and fallback to the "system" kubeconfig.

From a package point of view we need to default to getting the "system" kubeconfig as this is the expected behavior for CAPI & provider controllers. 

@ykakarap - i agree on the function signatures being a breaking API change and that we can just introduce a new function so as not to break any existing clients.

fabriziopandini commented on cmd/clusterctl/cmd/get_kubeconfig.go:
```diff
@@ -59,6 +63,8 @@ func init() {
 		"Path to the kubeconfig file to use for accessing the management cluster. If unspecified, default discovery rules apply.")
 	getKubeconfigCmd.Flags().StringVar(&gk.kubeconfigContext, "kubeconfig-context", "",
 		"Context to be used within the kubeconfig file. If empty, current context will be used.")
+	getKubeconfigCmd.Flags().BoolVar(&gk.userKubeconfig, "user", true,
+		"Fo providers that generates a user specific kubeconfig, it returns the user specific kubeconfig if set to true, else returns system generated kubeconfig.")
```
I think we should expand a little bit the semantics of this flag:
- by default it selects the kubeconfig for the users
- it allows to fetch the cluster kubeconfig when specifically required to

valaparthvi commented:
@fabriziopandini @richardcase @ykakarap 
I've modified the PR according to your changes.
1. Added a new method `GetUserKubeconfig` to `WorkloadCluster`.
2. Added a new function `FromUserSecret`
3. Added unit tests for both 1, and 2.
4. Modified `clusterClien.GetKubeconfig` method to use user kubeconfig by default, and unless that fails, or if system kubeconfig is explicitly asked for, then fetch system kubeconfig.

Also, my apologies for ghosting this PR for so long :raised_hands:

ykakarap commented on cmd/clusterctl/client/cluster/workload_cluster_test.go:
```diff
@@ -93,7 +96,87 @@ users:
+		validUserKubeconfig = `
+clusters:
+- cluster:
+    certificate-authority-data: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUN5RENDQWJDZ0F3SUJBZ0lCQURBTkJna3Foa2lHOXcwQkFRc0ZBREFWTVJNd0VRWURWUVFERXdwcmRXSmwKY201bGRHVnpNQjRYRFRFNU1ERXhNREU0TURBME1Gb1hEVEk1TURFd056RTRNREEwTUZvd0ZURVRNQkVHQTFVRQpBeE1LYTNWaVpYSnVaWFJsY3pDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBT1EvCmVndmViNk1qMHkzM3hSbGFjczd6OXE4QTNDajcrdnRrZ0pUSjRUSVB4TldRTEd0Q0dmL0x ...
```
Might be better to use a random string for the cert-authority-filed like the other test cases.

ykakarap commented on cmd/clusterctl/client/cluster/workload_cluster_test.go:
```diff
@@ -93,7 +96,87 @@ users:
+preferences: {}
+- name: test1-admin-user
+  user:
+    exec:
+      apiVersion: client.authentication.k8s.io/v1alpha1
+      args:
+      - token
+      - -i
+      - test1-admin-user
+      command: aws-iam-authenticator
+      env: null
+      provideClusterInfo: false
```
Since this is not relevant to the test itself can you please make the secret simpler. Using something similar to other example tests in the file should be good enough.

ykakarap commented on cmd/clusterctl/client/cluster/workload_cluster_test.go:
```diff
@@ -93,7 +96,87 @@ users:
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test1-user-kubeconfig",
+				Namespace: "test",
+				Labels:    map[string]string{clusterv1.ClusterLabelName: "test1-user"},
```
```suggestion
				Labels:    map[string]string{clusterv1.ClusterLabelName: "test1"},
```
To match with the naming convention of the user kubeconfig secret the cluster name should be `test1`, right?

valaparthvi replied:
I'd rather not, AWS appends `-user` to the  label as well.

ykakarap commented on cmd/clusterctl/client/cluster/workload_cluster_test.go:
```diff
@@ -93,7 +96,87 @@ users:
+		name           string
+		expectErr      bool
+		proxy          Proxy
+		userKubeConfig bool
```
Looks like this is unused.

ykakarap commented on cmd/clusterctl/client/cluster/workload_cluster_test.go:
```diff
@@ -93,7 +96,87 @@ users:
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			wc := newWorkloadCluster(tt.proxy)
+			data, err := GetUserKubeconfig(wc.proxy, "test1", "test")
```
```suggestion
			data, err := GetUserKubeconfig(tt.proxy, "test1", "test")
```

ykakarap commented on cmd/clusterctl/client/get_kubeconfig.go:
```diff
@@ -56,5 +61,15 @@ func (c *clusterctlClient) GetKubeconfig(options GetKubeconfigOptions) (string,
 	}
 
-	return clusterClient.WorkloadCluster().GetKubeconfig(options.WorkloadClusterName, options.Namespace)
+	// workload cluster kubeconfig is explicitly asked for
```
```suggestion
	// system kubeconfig is explicitly asked for
```

ykakarap commented on cmd/clusterctl/client/get_kubeconfig.go:
```diff
@@ -56,5 +61,15 @@ func (c *clusterctlClient) GetKubeconfig(options GetKubeconfigOptions) (string,
+	kubeconfig, err := cluster.GetUserKubeconfig(clusterClient.Proxy(), options.WorkloadClusterName, options.Namespace)
+	if err != nil {
+		// fallback on the workload cluster kubeconfig
+		klog.Warningf("failed to fetch the user kubeconfig; fetching the workload cluster kubeconfig instead")
```
```suggestion
		klog.Warningf("failed to fetch the user kubeconfig; fetching the system kubeconfig instead")
```

ykakarap commented on cmd/clusterctl/client/get_kubeconfig_test.go:
```diff
@@ -48,7 +48,7 @@ func Test_clusterctlClient_GetKubeconfig(t *testing.T) {
-			options:   GetKubeconfigOptions{Kubeconfig: Kubeconfig(kubeconfig)},
+			options:   GetKubeconfigOptions{Kubeconfig: Kubeconfig(kubeconfig), UserKubeconfig: true},
 			expectErr: true,
 		},
```
Can you also add a test case for the fallback flow?

valaparthvi replied:
I wanted to, but I couldn't figure out how to setup a fake cluster. I'll give it a shot again.

ykakarap replied:
I will take a closer look at this and drop something here in the coming days.

valaparthvi replied:
Thanks. There's no hurry at my end, I am traveling for the next 2 weeks, so I might not be able to work on this right away.

ykakarap commented on docs/book/src/clusterctl/provider-contract.md:
```diff
@@ -316,6 +316,11 @@ are set only for objects that are not directly referenced by Cluster API core ob
+### Secret naming convention
+Provider authors should be aware of the Kubeconfig secret naming conventions. If your provider generates a user specific Kubeconfig secret, the secret should have `<cluster-name>-user-kubeconfig` naming convention.
+
+`clusterctl get kubeconfig` will return this user specific Kubeconfig by default, and fallback to the CAPI generated Kubeconfig that follows a preset naming convention `<cluster-name>-kubeconfig`.
```
```suggestion
`clusterctl get kubeconfig` will return this user specific Kubeconfig by default, and fallback to the CAPI generated Kubeconfig that follows the `<cluster-name>-kubeconfig` naming convention.
```

ykakarap commented:
>Also, my apologies for ghosting this PR for so long 🙌

No problem 🙂 
Great to see that you are following up on the PR.

valaparthvi commented on cmd/clusterctl/client/client.go:
```diff
@@ -42,7 +42,7 @@ type Client interface {
 	GetClusterTemplate(options GetClusterTemplateOptions) (Template, error)
 
-	// GetKubeconfig returns the kubeconfig of the workload cluster.
+	// GetKubeconfig returns the user kubeconfig by default, but if it is unavailable or if explicitly asked for, then returns the workload cluster kubeconfig.
```
Honestly,I have been very confused with both the terms, but what you suggested here makes more sense.

ykakarap replied:
Can you also please make sure to make the appropriate changes in other parts of the PR?

valaparthvi commented on cmd/clusterctl/client/get_kubeconfig_test.go:
```diff
@@ -67,3 +75,70 @@ func Test_clusterctlClient_GetKubeconfig(t *testing.T) {
+    client-key-data: stuff-key-data
+`
+
+func Test_clusterctlClient_GetKubeconfig1(t *testing.T) {
```
@ykakarap I have added a temporary test case. It currently fails with the following error:
```sh
=== RUN   Test_clusterctlClient_GetKubeconfig1
    get_kubeconfig_test.go:142: 
        Unexpected error:
            <*errors.withStack | 0xc0002d63c0>: {
                error: <*errors.withMessage | 0xc0005ec080>{
                    cause: <*errors.StatusError | 0xc0003a50e0>{
                        ErrStatus: {
                            TypeMeta: {Kind: "", APIVersion: ""},
                            ListMeta: {
                                SelfLink: "",
                                ResourceVersion: "",
                                Continue: "",
                                RemainingItemCount: nil,
                            },
                            Status: "Failure",
                            Message: "customresourcedefinitions.apiextensions.k8s.io \"clusters.cluster.x-k8s.io\" not found",
                            Reason: "NotFound",
                            Details: {
                                Name: "clusters.cluster.x-k8s.io",
                                Group: "apiextensions.k8s.io",
                                Kind: "customresourcedefinitions",
                                UID: "",
                                Causes: nil,
                                RetryAfterSeconds: 0,
                            },
                            Code: 404,
                        },
                    },
                    msg: "failed to check Cluster API version",
                },
                stack: [0x182dee5, 0x18af66f, 0x18c7316, 0x18c729c, 0x516442, 0x46cf61],
            }
            failed to check Cluster API version: customresourcedefinitions.apiextensions.k8s.io "clusters.cluster.x-k8s.io" not found
        occurred
--- FAIL: Test_clusterctlClient_GetKubeconfig1 (0.00s)
```
While debugging this, I found out that `GetKubeconfig` calls `ProviderInvetory`, which uses the internalClient.proxy instead of the fakeProxy to obtain the objects (in this case the crd). The internal client of the clusterClient that is passed to `GetKubeconfig` during the test does not have any objects in it's proxy, hence the error.
https://github.com/kubernetes-sigs/cluster-api/blob/447bc5be5fb3a437962c1ce18080549612742883/cmd/clusterctl/client/get_kubeconfig.go#L44

The internalClient is created before any objects are added to the fake proxy. The solution that I see is either redefining the internalClient after objects are added, or modifying the `newFakeCluster` to take in an additional argument for proxy.
https://github.com/kubernetes-sigs/cluster-api/blob/447bc5be5fb3a437962c1ce18080549612742883/cmd/clusterctl/client/client_test.go#L208-L231

I cannot say for sure if there is something amiss with any of the helper functions here. Or perhaps I am using the wrong helper functions. What do you think?

valaparthvi replied:
I think I found out where I was going wrong. I will push the changes soon.

ykakarap commented on cmd/clusterctl/client/get_kubeconfig.go:
```diff
@@ -56,5 +61,15 @@ func (c *clusterctlClient) GetKubeconfig(options GetKubeconfigOptions) (string,
+	kubeconfig, err := cluster.GetUserKubeconfig(clusterClient.Proxy(), options.WorkloadClusterName, options.Namespace)
+	if err != nil {
+		// fallback on the workload cluster kubeconfig
+		klog.Warningf("failed to fetch the user kubeconfig; fetching the system kubeconfig instead")
```
Can you please switch to using the logger from `"sigs.k8s.io/cluster-api/cmd/clusterctl/log"`?

Ref for using it: https://github.com/ykakarap/cluster-api/blob/1da5d29c9ec54b9dcd6bb261b4cae8864a9d2734/cmd/clusterctl/client/init.go#L97

ykakarap commented on cmd/clusterctl/client/get_kubeconfig.go:
```diff
@@ -56,5 +61,15 @@ func (c *clusterctlClient) GetKubeconfig(options GetKubeconfigOptions) (string,
+	}
+	kubeconfig, err := cluster.GetUserKubeconfig(clusterClient.Proxy(), options.WorkloadClusterName, options.Namespace)
+	if err != nil {
+		// fallback on the workload cluster kubeconfig
```
```suggestion
		// fallback on the system kubeconfig
```

ykakarap commented on cmd/clusterctl/client/cluster/workload_cluster.go:
```diff
@@ -53,7 +54,30 @@ func (p *workloadCluster) GetKubeconfig(workloadClusterName string, namespace st
 	dataBytes, err := utilkubeconfig.FromSecret(ctx, cs, obj)
 	if err != nil {
-		return "", errors.Wrapf(err, "\"%s-kubeconfig\" not found in namespace %q", workloadClusterName, namespace)
+		errSecretName := utilsecret.Name(workloadClusterName, utilsecret.Kubeconfig)
+		return "", errors.Wrapf(err, "%q not found in namespace %q", errSecretName, namespace)
```
```suggestion
		secretName := utilsecret.Name(workloadClusterName, utilsecret.Kubeconfig)
		return "", errors.Wrapf(err, "%q not found in namespace %q", secretName, namespace)
```
nit

ykakarap commented on cmd/clusterctl/client/cluster/workload_cluster.go:
```diff
@@ -53,7 +54,30 @@ func (p *workloadCluster) GetKubeconfig(workloadClusterName string, namespace st
+	dataBytes, err := utilkubeconfig.FromUserSecret(ctx, cs, obj)
+
+	if err != nil {
+		errUserSecretName := utilsecret.Name(workloadClusterName, utilsecret.UserKubeconfig)
+		return "", errors.Wrapf(err, "%q not found in namespace %q", errUserSecretName, namespace)
```
```suggestion
		userSecretName := utilsecret.Name(workloadClusterName, utilsecret.UserKubeconfig)
		return "", errors.Wrapf(err, "%q not found in namespace %q", userSecretName, namespace)
```
nit

ykakarap commented on util/secret/consts.go:
```diff
@@ -32,6 +32,9 @@ const (
 	// Kubeconfig is the secret name suffix storing the Cluster Kubeconfig.
 	Kubeconfig = Purpose("kubeconfig")
 
+	// UserKubeconfig is the secret name suffix storing the Cluster Kubeconfig.
```
```suggestion
	// UserKubeconfig is the secret name suffix storing the Cluster's User Kubeconfig.
```

ykakarap commented on util/secret/consts.go:
```diff
@@ -32,6 +32,9 @@ const (
 	// Kubeconfig is the secret name suffix storing the Cluster Kubeconfig.
```
```suggestion
	// Kubeconfig is the secret name suffix storing the Cluster's System Kubeconfig.
```

ykakarap commented on util/secret/secret.go:
```diff
@@ -56,7 +56,12 @@ func Name(cluster string, suffix Purpose) string {
+	var separatorPos = strings.LastIndex(name, "-")
+
+	if strings.Contains(name, string(UserKubeconfig)) {
+		separatorPos = strings.LastIndexAny(name[:separatorPos], "-")
```
Out of curiosity: why use `strings.LastIndexAny` instead of `strings.LastIndex`?

ykakarap commented on cmd/clusterctl/client/get_kubeconfig_test.go:
```diff
@@ -48,9 +56,57 @@ func Test_clusterctlClient_GetKubeconfig(t *testing.T) {
+			options: GetKubeconfigOptions{
+				Kubeconfig:          Kubeconfig(kubeconfig),
+				Namespace:           "default",
+				WorkloadClusterName: "mgmt-cluster",
```
```suggestion
				WorkloadClusterName: "workload-cluster",
```
nit: Using `mgmt-cluster` as the name of the workload cluster could be misleading. 
(Please adjust at other places as well 🙂)

ykakarap reviewed (commented):
Great work on fixing the test cases 🚀 

The overall approach looks great. 

@valaparthvi Looking at the compatibility notice of clusterctl as a library (https://github.com/kubernetes-sigs/cluster-api/blob/main/cmd/clusterctl/README.md) we are not providing any guarantees at the moment about the API. While it will be nice to not make any big breaking changes looks like we can get a pass to make some as needed.

With this in mind, it might be cleaner to make the `GetUserKubeconfig` function part of the  `WorkloadCluster` interface instead of leaving it as a package level function. 

@fabriziopandini What are your thoughts on this?

valaparthvi commented:
> @valaparthvi Looking at the compatibility notice of clusterctl as a library (https://github.com/kubernetes-sigs/cluster-api/blob/main/cmd/clusterctl/README.md) we are not providing any guarantees at the moment about the API. While it will be nice to not make any big breaking changes looks like we can get a pass to make some as needed.
> 
> With this in mind, it might be cleaner to make the `GetUserKubeconfig` function part of the `WorkloadCluster` interface instead of leaving it as a package level function.

`GetUserKubeconfig` was explicitly defined as a function in the first place to avoid breaking the API.(https://github.com/kubernetes-sigs/cluster-api/pull/6346#discussion_r857891414)

ykakarap commented:
> > @valaparthvi Looking at the compatibility notice of clusterctl as a library (https://github.com/kubernetes-sigs/cluster-api/blob/main/cmd/clusterctl/README.md) we are not providing any guarantees at the moment about the API. While it will be nice to not make any big breaking changes looks like we can get a pass to make some as needed.
> > With this in mind, it might be cleaner to make the `GetUserKubeconfig` function part of the `WorkloadCluster` interface instead of leaving it as a package level function.
> 
> `GetUserKubeconfig` was explicitly defined as a function in the first place to avoid breaking the API.([#6346 (comment)](https://github.com/kubernetes-sigs/cluster-api/pull/6346#discussion_r857891414)) But, if it makes sense, I can implement this as a method.

I would suggest that you wait on implementing this till we get confirmation on a direction to take :)

valaparthvi commented:
> > > @valaparthvi Looking at the compatibility notice of clusterctl as a library (https://github.com/kubernetes-sigs/cluster-api/blob/main/cmd/clusterctl/README.md) we are not providing any guarantees at the moment about the API. While it will be nice to not make any big breaking changes looks like we can get a pass to make some as needed.
> > > With this in mind, it might be cleaner to make the `GetUserKubeconfig` function part of the `WorkloadCluster` interface instead of leaving it as a package level function.
> > 
> > 
> > `GetUserKubeconfig` was explicitly defined as a function in the first place to avoid breaking the API.([#6346 (comment)](https://github.com/kubernetes-sigs/cluster-api/pull/6346#discussion_r857891414)) But, if it makes sense, I can implement this as a method.
> 
> I would suggest that you wait on implementing this till we get confirmation on a direction to take :)

Can I take that up in another PR then? Or would you rather I do it in the same one?

sbueringer commented:
Let's change the func to a method please (and in this PR directly). As Yuvaraj wrote we have no strong API guarantees on this package and making it a func makes the API in my opinion very inconsistent.

valaparthvi commented:
How do I fix the apidiff failures?

sbueringer commented:
> How do I fix the apidiff failures?

You don't have to. It just indicates to us that there's an API change and we can decide to accept it (it's optional for merge)

sbueringer commented on cmd/clusterctl/client/cluster/workload_cluster.go:
```diff
@@ -53,7 +55,30 @@ func (p *workloadCluster) GetKubeconfig(workloadClusterName string, namespace st
 }
+
+// GetUserKubeconfig returns the kubeconfig of the user as provided by the provider.
+// Note: This has not been converted to a method yet to prevent API change.
```
```suggestion
```

sbueringer commented on cmd/clusterctl/client/cluster/workload_cluster.go:
```diff
@@ -53,7 +55,30 @@ func (p *workloadCluster) GetKubeconfig(workloadClusterName string, namespace st
 	if err != nil {
-		return "", errors.Wrapf(err, "\"%s-kubeconfig\" not found in namespace %q", workloadClusterName, namespace)
+		secretName := utilsecret.Name(workloadClusterName, utilsecret.Kubeconfig)
+		return "", errors.Wrapf(err, "%q not found in namespace %q", secretName, namespace)
```
Let's just generate the error inside of the FromSecret func

same for FromUserSecret

sbueringer commented on cmd/clusterctl/client/get_kubeconfig.go:
```diff
@@ -56,5 +61,16 @@ func (c *clusterctlClient) GetKubeconfig(options GetKubeconfigOptions) (string,
+	if err != nil {
+		// fallback on the system kubeconfig
+		logger := log.Log
+		logger.Info("failed to fetch the user kubeconfig; fetching the system kubeconfig instead")
```
Please only return an error here don't log + return an error

+ wrap the error as suggested in the other comments

ykakarap replied:
How about doing error check on the call to get the system kubeconfig and log only if are able to get the system kubeconfig?

sbueringer replied:
Why do we need it? I would assume:
* no error => we just print the kubeconfig
* error => we print the kubeconfig

sbueringer commented on cmd/clusterctl/client/get_kubeconfig.go:
```diff
@@ -56,5 +61,16 @@ func (c *clusterctlClient) GetKubeconfig(options GetKubeconfigOptions) (string,
+		logger.Info("failed to fetch the user kubeconfig; fetching the system kubeconfig instead")
+		return clusterClient.WorkloadCluster().GetKubeconfig(options.WorkloadClusterName, options.Namespace)
+	}
+	return kubeconfig, err
```
```suggestion
	return kubeconfig, nil
```

sbueringer commented on cmd/clusterctl/cmd/get_kubeconfig.go:
```diff
@@ -40,11 +41,14 @@ var getKubeconfigCmd = &cobra.Command{
 
 	Example: Examples(`
-		# Get the workload cluster's kubeconfig.
+		# Get the workload cluster's user kubeconfig. If it is not available, then the workload cluster's system kubeconfig is returned.
```
Is it intentional that we change the default behavior of get kubeconfig?

sbueringer replied:
I think the fallback is not implemented?

ykakarap replied:
It was previously discussed top make the default behavior to get the user kubeconfig. If the user config is not available then the command should download the workload cluster's system kubeconfig. Ref previous discussion here: https://github.com/kubernetes-sigs/cluster-api/issues/3661#issuecomment-694747492

And looke like the fallback is already implemented.

sbueringer replied:
Where is the fallback implemented? Not sure how I'm missing it but I went over the PR again, I can't find it.

Are we sure this automagic fallback is a good idea? Essentially this means you run a clusterctl command and can never be sure if you end up with a user or system kubeconfig.

sbueringer replied:
To be clear with fallback I understand: `clusterctl get kubeconfig` is run and it will return if it exists the user kubeconfig and if it doesn't exist the system kubeconfig.

sbueringer replied:
But in my opinion the current implementation as I understand it (no automatically returning a different kubeconfig if no user kubeconfig exists) is what we should do, the comment just doesn't reflect it.

sbueringer replied:
Found it ...

Do we really think this is a good idea? 

If I want to get a user kubeconfig and it doesn't exist I will just get the system kubeconfig automatically. That sounds pretty dangerous. I think we can't rely on some output signalling it given that `clusterctl get kubeconfig` is likely to be also run inside of some pipelines where nobody will notice a log line that clusterctl just returned another kubeconfig.

sbueringer replied:
Just discussed it with Yuvaraj, we would want to discuss this with Fabrizio first (he will be back next week)

valaparthvi replied:
Hi @fabriziopandini, looping you into this thread since we're waiting on you to finalize the behavior on this command.

valaparthvi replied:
> Found it ...
> 
> I'm not sure if it is a good idea to be honest
> 
> If I want to get a user kubeconfig and it doesn't exist (or doesn't exist yet) I will just get the system kubeconfig automatically. That sounds pretty dangerous (I'm assuming a user kubeconfig doesn't have the same full access as the system kubeconfig). I think we can't rely on some output signalling it given that `clusterctl get kubeconfig` is likely to be also run inside of some pipelines where nobody will notice a log line that clusterctl just returned another kubeconfig.

@fabriziopandini Do you think we should change the default behavior?

richardcase replied:
@sbueringer - so are you thinking if you request the user kubeconfig and it doesn't exist (and perhaps never will for a specific provider) that it returns an error message / specific exit code?

fabriziopandini replied:
I think that Stefan expressed a valid concern, given that falling back from a kubeconfig to another can lead to security issues (if the two kubeconfig have higher RBAC premissions)

A possible way forward is that `clusterctl get kubeconfig` could support retrieving a different kubeconfig only if this is specifically requested, e.g. `clusterctl get kubeconfig --type=user`.

However, I have some concerns about this approach as well because:
- It requires providing EKS-specific instructions to the users, and with this regards it is not much different than the current situation where we are giving instruction to retrieve the user kubeconfig via clusterctl
- It could give users the impression that clusterctl is a tool to manage many kubeconfig, which is not something we should do IMO.

The only alternative is to define a strict contract for providers (and BTW, no matter the decision I do expect this PR to extend the contract for control plane providers) that requires that control plane providers should create kubeconfig in a well-defined order (first user-kubeconfig, the defualt kubeconfig).
TBH I do not know if this is acceptable as idea given that there is still a chance of privilege escalation if  for any reason the user kubeconfig gets deleted

fabriziopandini replied:
Discussed in Aug 3rd Office hours
We are open to idea on how to make UX nice but we cannot trade on security concern
Also, it will be great if we can discuss this in terms of contract all the control plane providers should abide to.

ykakarap commented on cmd/clusterctl/client/get_kubeconfig.go:
```diff
@@ -56,5 +61,16 @@ func (c *clusterctlClient) GetKubeconfig(options GetKubeconfigOptions) (string,
+	if !options.UserKubeconfig {
+		return clusterClient.WorkloadCluster().GetKubeconfig(options.WorkloadClusterName, options.Namespace)
+	}
+	kubeconfig, err := clusterClient.WorkloadCluster().GetUserKubeconfig(options.WorkloadClusterName, options.Namespace)
```
```suggestion
    // fetch the workload cluster's user kubeconfig.
	kubeconfig, err := clusterClient.WorkloadCluster().GetUserKubeconfig(options.WorkloadClusterName, options.Namespace)
```
nit

ykakarap commented on cmd/clusterctl/client/cluster/workload_cluster.go:
```diff
@@ -21,12 +21,14 @@ import (
 type WorkloadCluster interface {
 	// GetKubeconfig returns the kubeconfig of the workload cluster.
 	GetKubeconfig(workloadClusterName string, namespace string) (string, error)
+	GetUserKubeconfig(workloadClusterName string, namespace string) (string, error)
```
```suggestion

    // GetUserKubeconfig returns the system kubeconfig of the  workload cluster.
	GetUserKubeconfig(workloadClusterName string, namespace string) (string, error)
```

sbueringer replied:
"GetUserKubeconfig returns the system" doesn't sound right

ykakarap commented on cmd/clusterctl/client/cluster/workload_cluster.go:
```diff
@@ -21,12 +21,14 @@ import (
 
 // WorkloadCluster has methods for fetching kubeconfig of workload cluster from management cluster.
 type WorkloadCluster interface {
 	// GetKubeconfig returns the kubeconfig of the workload cluster.
```
```suggestion
	// GetKubeconfig returns the system kubeconfig of the workload cluster.
```

ykakarap commented on cmd/clusterctl/client/get_kubeconfig_test.go:
```diff
@@ -63,7 +120,62 @@ func Test_clusterctlClient_GetKubeconfig(t *testing.T) {
 			}
 			g.Expect(err).ToNot(HaveOccurred())
-			g.Expect(config).ToNot(BeEmpty())
+			g.Expect(config).To(BeEquivalentTo(tt.want))
```
```suggestion
			g.Expect(config).To(Equal(tt.want))
```
since actual and expected are both of the same type (string) we can use the `Equal` matcher here.

ykakarap commented on cmd/clusterctl/client/get_kubeconfig_test.go:
```diff
@@ -17,40 +17,97 @@ limitations under the License.
+			want:      getKubeconfig(clusterName, false),
+		},
+		{
+			name: "returns the system kubeconfig when GetKubeconfigOptions.UserKubeconfig is set to false",
```
Can you please add both user kubeconfig and system kubeconfig to the list of fake object. When the `UserKubeconfig` is set to true we want the function to return the system kubeconfig even if the user kubeconfig is present.

valaparthvi commented:
> Just discussed it with Yuvaraj, we would want to discuss this with Fabrizio first (he will be back next week)

I'm going to wait on the requested changes until we finalize the behavior for this command.

valaparthvi commented:
@ykakarap I've worked on the changes you requested.

valaparthvi commented:
> Discussed in Aug 3rd Office hours We are open to idea on how to make UX nice but we cannot trade on security concern Also, it will be great if we can discuss this in terms of contract all the control plane providers should abide to.
https://github.com/kubernetes-sigs/cluster-api/pull/6346#discussion_r936949742

@richardcase In response to the above comment by @fabriziopandini, I think I should close this PR until the terms of contract for all the control plane providers are finalized. Let me know if you think there should be another course of action, or if I can help with that.

richardcase commented:
> > Discussed in Aug 3rd Office hours We are open to idea on how to make UX nice but we cannot trade on security concern Also, it will be great if we can discuss this in terms of contract all the control plane providers should abide to.
> > [#6346 (comment)](https://github.com/kubernetes-sigs/cluster-api/pull/6346#discussion_r936949742)
> 
> @richardcase In response to the above comment by @fabriziopandini, I think I should close this PR until the terms of contract for all the control plane providers are finalized. Let me know if you think there should be another course of action, or if I can help with that.

I get the concern around falling back and inadvertently getting a higher privileged kubeconfig. For me, this is around enabling a cluster operator to easily get a kubeconfig that can be given to users...and trying to have consistency in how its retrieved across providers. As an operator i'd hate to use 1 method for CAPA and another method for a different provider.

I like the idea of adding `--type=user` or something similar...but with no fallback. 

But overall, what I am hearing from the comments is that more thought & design is needed for this change to ensure we cover all concerns raised....so i'm thinking we create a proposal for this.

@valaparthvi  - did you want to work together on a proposal for this change?

@fabriziopandini - do you think doing a proposal would help?

valaparthvi commented:
>@valaparthvi - did you want to work together on a proposal for this change?

I would love to, but I'll need a lot of guidance on how to go forward with the proposal.
